go/cmd/opensearch/main.go

```diff
@@ -0,0 +1,656 @@
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	b, err := bgzf.NewReader(tarReader, 0)
+	if err != nil {
+		archive.Seek(0, 0)
+
+		tarReader, fileStats, err := _getAnnotationFhFromTarArchive(archive)
+
+		if err != nil {
+			return nil, nil, err
+		}
+
+		b, err := gzip.NewReader(tarReader)
+
+		if err != nil {
+			return nil, nil, err
+		}
+
+		bufioReader := bufio.NewReader(b)
+
+		return BufioBystroReader{Reader: bufioReader}, fileStats, nil
+	}
+
+	return BzfBystroReader{Reader: b}, fileStats, err
+}
+
+func sendEvent(message ProgressMessage, eventTube *beanstalk.Tube, noBeanstalkd bool) {
+	if noBeanstalkd {
+		fmt.Printf("Indexed %d annotated variants\n", message.Data.Progress)
+		return
+	}
+
+	messageJson, err := sonic.Marshal(message)
+	if err != nil {
+		log.Fatalf("Marshaling failed of progress message: [%s]\n", err)
+	}
+
+	eventTube.Put(messageJson, 0, 0, 0)
+}
+
+func main() {
+	cliargs := setup(nil)
+
+	indexName := cliargs.indexName
+
+	// From testing, performance seems maximized at 32 threads for a 4 vCPU AWS instance
+	concurrency := runtime.NumCPU() * 8
+
+	workQueue := make(chan parser.Job)
+	complete := make(chan bool)
+
+	archive, err := os.Open(cliargs.annotationTarballPath)
+	if err != nil {
+		log.Fatalf("Couldn't open tarball due to: [%s]\n", err)
+	}
+	defer archive.Close()
+
+	reader, fileStats, err := getAnnotationFhFromTarArchive(archive)
+
+	if err != nil {
+		log.Fatalf("Couldn't get annotation file handle due to: [%s]\n", err)
+	}
+
+	headerPaths, headerFields := getHeaderPaths(reader)
+
+	if len(headerPaths) == 0 {
+		log.Fatal("No header found")
+	}
+
+	osConfig, client, osearchMapConfig := createIndex(cliargs.osConnectionConfigPath, cliargs.osIndexConfigPath, cliargs.indexName, fileStats.Size())
+
+	// Spawn threads
+	for i := 0; i < concurrency; i++ {
+		go parser.Parse(headerPaths, indexName, osConfig, workQueue, complete, i)
+	}
+
+	var beanstalkdConfig BeanstalkdConfig
+	var eventTube *beanstalk.Tube
+	var message ProgressMessage
+
+	if !cliargs.noBeanstalkd {
+		beanstalkdConfig, err = createBeanstalkdConfig(cliargs.beanstalkConfigPath)
+		if err != nil {
+			log.Fatalf("Couldn't create beanstalkd config due to: [%s]\n", err)
+		}
+		beanstalkConnection, err := beanstalk.Dial("tcp", beanstalkdConfig.Addresses[0])
+		if err != nil {
+			log.Fatalf("Couldn't connect to beanstalkd due to: [%s]\n", err)
+		}
+		defer beanstalkConnection.Close()
+		eventTube = beanstalk.NewTube(beanstalkConnection, beanstalkdConfig.Tubes.Index.Events)
+
+		message = ProgressMessage{
+			SubmissionID: cliargs.jobSubmissionID,
+			Event:        PROGRESS_EVENT,
+			Data: ProgressData{
+				Progress: 0,
+				Skipped:  0,
+			},
+		}
+	}
+
+	progressUpdate := 0
+	chunkStart := 0
+	var lines []string
+
+	for {
+		rawLines, err := reader.readLines()
+
+		if len(rawLines) > 0 {
+			lines = strings.Split(string(rawLines), parser.LINE_DELIMITER)
+		}
+
+		if err != nil {
```

Comment:
1. Typically, you'd check for an error before doing anything with returned data. 
2. Ultimately, you read a chunk of lines and then split those lines. It would seem more straightforward to use a bufio.Scanner that is reading from either the gzip or bgzip reader interface. If you want to enlarge the buffer to improve performance, you can call scanner.Buffer() to set the byte slice and max size. The default scanner breaks on lines so you could either accumulate lines or hand off individual lines to the parser. Did you try this approach and find it was problematic?

Reply:
I've created a ticket to evaluate this approach. bgzf's use of transactions would I believe require me to wrap the scanner call in one long transaction, but I don't know how problematic that is.